backend/src/utils/chat.ts

```diff
@@ -11,7 +11,10 @@ function pushMessageToHistory(
 	// remove the oldest message, not including system role message
 	// until the length of the chat history is less than maxChatHistoryLength
 	while (updatedChatHistory.length >= maxChatHistoryLength) {
-		if (updatedChatHistory[0].completion?.role !== 'system') {
+		if (
+			'completion' in updatedChatHistory[0] &&
+			updatedChatHistory[0].completion?.role !== 'system'
+		) {
 			updatedChatHistory.shift();
```

Comment:
I believe we always expect the system role message at the very start of the chat history (or not at all). If so, it'd be more efficient to do something like this:

```
const spliceFrom = 
  'completion' in updatedChatHistory[0] && updatedChatHistory[0].completion?.role === 'system'
    ? 1
    : 0;

while (updatedChatHistory.length >= maxChatHistoryLength) {
  updatedChatHistory.splice(spliceFrom, 1);
}
```

Unsure if you can use `instanceof` instead of `in`, I have a feeling that only works with classes and interfaces:

```
const spliceFrom = updatedChatHistory[0] instanceof ChatSystemMessage ? 1 : 0;
```

Reply:
I've refactored the whole lot to something that (I think) is more straightforward. Let me know what you think